designs/helper_to_upgrade_projects_by_rescaffolding.md

```diff
@@ -0,0 +1,155 @@
+
+```shell
+kubebuilder alpha generate [OPTIONS]
+```
+
+## Open Questions
+
+N/A
+
+## Summary
+
+The [PROJECT config][project-config] tracks all configuration and inputs used to create the
+project with the CLI. Therefore, a command could be able to read this information and re-run
+the same commands with the same data used previously to re-generate the project.
+
+## Motivation
+
+The recommended straightforward steps to upgrade the projects are re-scaffold all from scract using the upper
+version/plugin and then, re-add all code source on top again. Therefore, this command would help users in
+order to minimize the required effort by automating which shows technically possible of this process.
+
+The main motivation of this proposal is to provide a helper for upgrades and
+make less painful this process. Examples:
+
+- See the discussion [How to regenerate scaffolding?](https://github.com/kubernetes-sigs/kubebuilder/discussions/2864)
+- From [slack channel By Paul Laffitte](https://kubernetes.slack.com/archives/CAR30FCJZ/p1675166014762669)
+
+### Goals
+
+- Help users upgrade their project with the latest changes
+- Help users to re-scaffold the projects from the scratch based on what was done previously with the tool
+- Make less painful the process to upgrade
+
+### Non-Goals
+
+- Change the default layout or how the KubeBuilder CLI works
+- Deal with customizations or deviations from the proposed layout
+- Be able to perform the project upgrade to the latest changes without human bean interactions
+- Deal and support external plugins
+- Provides support to [declarative](https://book.kubebuilder.io/plugins/declarative-v1.html) plugin
+  since it is desired and planned to decouple this solution and donate this plugin to its own authors [More info](https://github.com/kubernetes-sigs/kubebuilder/issues/3186)
+- Provide support old version prior we have the Project config (Kubebuilder < 3x) and the go/v2 layout which exist to ensure  a backwards compatibility with legacy layout provide by Kubebuilder 2x
+
+## Proposal
+
+The proposed solution to achieve this goal is to create an alpha command as described
+in the example section above, see:
+
+```shell
+kubebuilder alpha generate \
+    --from-project-file=<path> 
+    --to=<path where the project should be re-scaffold> 
+    --backup=<path-where the current version of the project should be copied as backup>
+    --layout-plugins=<chain of plugins key that can be used to create the layout with init sub-command>
+```
+
+Where:
+
+- project-file: If not informed then, the command would check it in the current directory
+- output: If not informed then, it should be the current repository
+- backup: if not informed then, it would be the current path -1 level with the name backup
+- layout-plugins: if not informed then, it is the same plugin chain available in the layout field
+
+This command would mainly perform the following operations:
+
+- 1. Make a backup of the current project
```

Comment:
I'm curious about the reason to have the backup process here.

Say the author will re-scaffold the project in a new place, where the origin project will not be affected, then what is the use case to back up the origin?

Reply:
Because if they are not using git, they want to use the local backup to compare the scaffold done from scratch, etc. We can indeed improve that and add a flag where the backup would only generate if/when that is informed.

Reply:
I added it as a user story as well. Good point.